### PR TITLE
feat: whitelist some commands

### PR DESCRIPTION
**Pull request by brayo-pip**
> [!IMPORTANT]
> Adds a whitelist for shell commands in `execute_shell()` to restrict execution and require confirmation for non-whitelisted commands.
> 
>   - **Security**:
>     - Adds a whitelist for shell commands in `execute_shell()` to restrict execution to `ls`, `stat`, `cd`, `cat`, `pwd`, `echo`.
>     - Uses regex to parse commands and checks against the whitelist.
>     - Requires user confirmation for non-whitelisted commands if `ask` is true.
>   - **Misc**:
>     - Updates message formatting in `execute_shell()` to indicate if a command is whitelisted.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ErikBjare%2Fgptme&utm_source=github&utm_medium=referral)<sup> for c3fe3343da337c99c41925e93b9b010fd70346d7. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by 0xbrayo** [comment]
@ErikBjare Do I need to configure something on my end to fix these tests? I see one needs a `OPENAI_KEY`. They seem to be working on the other PRs though.

**Comment by ErikBjare** [comment]
Ah yes, since the keys are secret you don't have access to them when you trigger the job. Not sure what's the easiest way to work around this.

The logs contain `Secret source: None`

Weird that it worked for other PRs? It kinda shouldn't?

**Comment by ErikBjare** [comment]
I don't think it worked in other PRs either:

 - https://github.com/ErikBjare/gptme/pull/171
 - https://github.com/ErikBjare/gptme/pull/141

**Comment by ErikBjare** [comment]
> Anyone with collaborator access to this repository can use these secrets and variables for actions. They are not passed to workflows that are triggered by a pull request from a fork. 

I added you as a collaborator (invite sent), but you might have to set the secrets in your forked repo.

**Comment by 0xbrayo** [comment]
> @ErikBjare Do I need to configure something on my end to fix these tests? I see one needs a `OPENAI_KEY`. They seem to be working on the other PRs though.

Sorry I meant it seems to be working on your PRs

**Comment by 0xbrayo** [comment]
@ErikBjare I set the both openai and anthropic api keys in my repo secrets. But the errors are still there :(

**Comment by 0xbrayo** [comment]
@ellipsis-dev review

**Comment by 0xbrayo** [comment]
@ErikBjare I think this is ready, do I merge? The list of whitelisted commands could be extended but that can be done later.

**Comment by ErikBjare** [comment]
Sure, but ask for a review first next time :)

File: gptme/tools/shell.py [hunk 1]
```diff
@@ -243,15 +243,27 @@ def execute_shell(
...
     cmd = code.strip()
     if cmd.startswith("$ "):
         cmd = cmd[len("$ ") :]
 
-    print_preview(cmd, "bash")
-    if not confirm("Run command?"):
-        yield Message("system", "Command not run")
-        return
+    #NOTE: This does not handle control flow words like if, for, while.
+    regex = r"(?:^|[|&;]|\|\||&&|\n)\s*([^\s|&;]+)"
```
Comment (0xbrayo) [inline gptme/tools/shell.py:254]: I tested the regex extensively but I guess it's the most unpredictable bit here.

Comment (0xbrayo) [reply gptme/tools/shell.py:254]: [regex testing site](https://regex101.com/r/1e57VF/1)

File: gptme/tools/shell.py [hunk 2]
```diff
@@ -243,15 +243,27 @@ def execute_shell(
...
+
+    for match in re.finditer(regex, cmd):
```
Comment (ErikBjare) [inline gptme/tools/shell.py:256]: You should compile this regex

**Pull request merged**